### PR TITLE
Fix rabbitmq_policy tags json.loads - Issue #28

### PR DESCRIPTION
**Pull request by nazunalika**
Redoing PR based on feedback in #58

##### SUMMARY
Fixes #28 by trying to load with `json.loads` and if unable, to assign `tags` without loading as json.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
`rabbitmq_policy`

##### ADDITIONAL INFORMATION
During the `_policy_check` piece of this module, `policy_data` is basically a list based on a split of the variable `policy`. However, `policy` does not contain data. As a result of the split, it becomes: 

`['vhost', 'name', 'pattern', 'apply-to', 'definition', 'priority']`

As there is no JSON data, there is no reason to use `json.loads` on a string.

To reproduce this issue:
1. Install the ansible collection using ansible-galaxy
2. Create a playbook with a set of tasks that deal with rabbitmq policies on a vhost. This example is on a Red Hat system. Example tasks below.

```
- name: Configure pubsub virtual host
  community.rabbitmq.rabbitmq_vhost:
    name: /pubsub
    state: present

- name: Configure an HA policy for pubsub
  community.rabbitmq.rabbitmq_policy:
    name: HA
    apply_to: queues
    pattern: ".*"
    tags:
      ha-mode: all
      ha-sync-mode: automatic
      ha-sync-batch-size: 10000
    vhost: /pubsub
```

3. Run the playbook

```
# BEFORE
The full traceback is:
Traceback (most recent call last):
  File "/home/ansible/.ansible/tmp/ansible-tmp-1611654783.6917102-146428-116716834350778/AnsiballZ_rabbitmq_policy.py", line 102, in <module>
    _ansiballz_main()
  File "/home/ansible/.ansible/tmp/ansible-tmp-1611654783.6917102-146428-116716834350778/AnsiballZ_rabbitmq_policy.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/ansible/.ansible/tmp/ansible-tmp-1611654783.6917102-146428-116716834350778/AnsiballZ_rabbitmq_policy.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.community.rabbitmq.plugins.modules.rabbitmq_policy', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_community.rabbitmq.rabbitmq_policy_payload__4toqeqo/ansible_community.rabbitmq.rabbitmq_policy_payload.zip/ansible_collections/community/rabbitmq/plugins/modules/rabbitmq_policy.py", line 248, in <module>
  File "/tmp/ansible_community.rabbitmq.rabbitmq_policy_payload__4toqeqo/ansible_community.rabbitmq.rabbitmq_policy_payload.zip/ansible_collections/community/rabbitmq/plugins/modules/rabbitmq_policy.py", line 237, in main
  File "/tmp/ansible_community.rabbitmq.rabbitmq_policy_payload__4toqeqo/ansible_community.rabbitmq.rabbitmq_policy_payload.zip/ansible_collections/community/rabbitmq/plugins/modules/rabbitmq_policy.py", line 161, in has_modifications
  File "/tmp/ansible_community.rabbitmq.rabbitmq_policy_payload__4toqeqo/ansible_community.rabbitmq.rabbitmq_policy_payload.zip/ansible_collections/community/rabbitmq/plugins/modules/rabbitmq_policy.py", line 161, in <genexpr>
  File "/tmp/ansible_community.rabbitmq.rabbitmq_policy_payload__4toqeqo/ansible_community.rabbitmq.rabbitmq_policy_payload.zip/ansible_collections/community/rabbitmq/plugins/modules/rabbitmq_policy.py", line 198, in _policy_check
  File "/usr/lib64/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

#AFTER
TASK [rockylinux.rabbitmq : Configure the HA policy for pubsub] *********************************************
changed: [rabbitmq001.example.com]
```

### CODE REVIEW

**Comment by Andersson007** [comment]
@nazunalika hi, thanks for the new PR!
Strange errors.. As i can see there can be some issues with this https://stackoverflow.com/questions/15170676/cannot-import-jsondecodeerror.
Let's try to change `JSONDecodeError` to `ValueError`. If it doesn't help, to just `except Exception`?
I think it should be also fine.

**Comment by Andersson007** [comment]
Ah, i see now.. First please try to put `json.decoder.JSONDecodeError` instead `JSONDecodeError`, then, if doesn't help, what i suggested above. Thanks!

File: plugins/modules/rabbitmq_policy.py [hunk 1]
```diff
@@ -195,7 +195,12 @@ def _policy_check(self,
         policy_name = policy_data[name_fno]
         apply_to = policy_data[apply_to_fno]
         pattern = policy_data[pattern_fno].replace('\\\\', '\\')
-        tags = json.loads(policy_data[tags_fno])
+
+        try:
+            tags = json.loads(policy_data[tags_fno])
+        except JSONDecodeError:
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_policy.py:201]:
```suggestion
        except json.decoder.JSONDecodeError:
```

**Comment by Andersson007** [comment]
@nazunalika thanks! Green now!
Does this solve your problem locally? If yes, can we merge the PR?

File: changelogs/fragments/86-fix-tags-json-in-rmq-policy.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+  - >
+    rabbitmq_policy - The `_policy_check` piece of the policy module (`policy_data`) is typically list based on a split of the variable `policy`. However `policy` in some cases does not contain data. The fix allows `tags` to attempt to load as json first but in the case of failure, assign `tags` without using the json loader.
```
Comment (Andersson007) [inline changelogs/fragments/86-fix-tags-json-in-rmq-policy.yml:3]:
```suggestion
  - rabbitmq_policy - The ``_policy_check`` piece of the policy module (``policy_data``) is typically list based on a split of the variable ``policy``. However ``policy`` in some cases does not contain data. The fix allows ``tags`` to attempt to load as json first but in the case of failure, assign ``tags`` without using the json loader (https://github.com/ansible-collections/community.rabbitmq/pull/28).
```

**Comment by Andersson007** [comment]
cc @odyssey4me

**Review by odyssey4me — APPROVED** [review]
Nice, seems like a good way to get the best of both worlds. Thanks!

**Comment by Andersson007** [comment]
@nazunalika thanks for your contribution (bugfix + tests)! I think now we can release the collection even tomorrow, not on Friday as was planned.
Please take a look at the [open issues](https://github.com/ansible-collections/community.rabbitmq/issues) and [pull requests](https://github.com/ansible-collections/community.rabbitmq/pulls). Any help will be much appreciated!

@odyssey4me thanks for reviewing and approving!

**Pull request merged**